Update extension documentation

aledeg opened the pull request:
Changes proposed in this pull request:

- update extension documentation for the extension system

Pull request checklist:

- [x] clear commit messages
- [ ] code manually tested
- [ ] unit tests written (optional if too hard)
- [x] documentation updated

[Additional information can be found in the documentation](https://github.com/FreshRSS/FreshRSS/tree/master/docs/en/developers/04_Pull_requests.md).

aledeg commented:
@kevinpapst Do you have some thoughts about that?

Frenzie reviewed (commented):
For most (all?) of these it's not quite clear to me why they shouldn't be, even if on average it might be a bit silly to. Perhaps there's a distinction here between the `register*` methods and the rest.

aledeg commented:
I know! That's why I am asking for help here. But to me either the doc is wrong or the code is wrong, or worst both of them. And no matter what we choose, we have to fix something. I've tried with the code because it was faster but I need input on that to handle that properly.

Frenzie commented:
I'm not sure which doc you're referring to exactly, but my vote is probably
against the doc.

On Sat, Jan 9, 2021 at 4:54 PM Alexis Degrugillier <notifications@github.com>
wrote:

> I know! That's why I am asking for help here. But to me either the doc is
> wrong or the code is wrong, or worst both of them. And no matter what we
> choose, we have to fix something. I've tried with the code because it was
> faster but I need input on that to handle that properly.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/FreshRSS/FreshRSS/pull/3352#issuecomment-757326309>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABRQBPXUOALMWR3NOLHE7LSZB345ANCNFSM4V3VM2NQ>
> .
>

aledeg commented:
This one → https://github.com/FreshRSS/FreshRSS/blob/master/docs/en/developers/03_Backend/05_Extensions.md#writing-your-own-extensionphp

Frenzie commented:
I'd say that should be understood along [these lines](https://tools.ietf.org/html/rfc2119). That is, as synonymous with not recommended but you may still have a reason to. On average it doesn't strike me as something that should be enforced.

> 1. MUST   This word, or the terms "REQUIRED" or "SHALL", mean that the
   definition is an absolute requirement of the specification.
>
> 2. MUST NOT   This phrase, or the phrase "SHALL NOT", mean that the
   definition is an absolute prohibition of the specification.
> 
> 3. SHOULD   This word, or the adjective "RECOMMENDED", mean that there
   may exist valid reasons in particular circumstances to ignore a
   particular item, but the full implications must be understood and
   carefully weighed before choosing a different course.
>
> 4. SHOULD NOT   This phrase, or the phrase "NOT RECOMMENDED" mean that
   there may exist valid reasons in particular circumstances when the
   particular behavior is acceptable or even useful, but the full
   implications should be understood and the case carefully weighed
   before implementing any behavior described with this label.

kevinpapst commented:
Setting public methods in an extension point to final is a brutal BC break.
I don't see enough evidence for `should be left untouched` in 50% of the methods. The other ones are simply mislocated in that class.

Just a hypothetical example: a plugin needs the xyz extension. It checks for its existence in `enable()` before it activates itself. By setting `enable()` to final you break that plugin during the next update.
And now the users FRSS installation is broken 😢 

I am totally on your side regarding some of the methods, which shouldn't be in the users hand.
Maybe you can reach your goal in another way that cannot break an installation? Lets take `registerController()` for example. Instead of setting it final you remove it from the class and put the logic into the appropriate manager/service/whatever code.

So wherever `registerController()` was called before with the $base_name parameter you change:

Before:
```
$plugin->registerController($base_name);
```

After:
```
if (method_exists('registerController', $plugin)) {
    @trigger_error(sprintf('Implementing registerController() in your plugin %s is deprecated. Please update the plugin'), get_class($plugin)), E_USER_DEPRECATED);
}
Minz_Dispatcher::registerController($base_name, $plugin->getPath());
```

That is still a potential BC break in the plugin, but should at least not break the complete application.

But again (like the last PR) I am not sure what you gain by these changes? Is there a roadmap for changes in the future that require them? Sometimes it is just too late to change some code... no matter how deeply your developer hard bleeds 😁

aledeg commented:
@kevinpapst Thank you for your thoughts.
My goal here is to have something that is correct. What I mean by that is that the documentation is in sync with the code and vice-versa. Here is not the case.
As I mentioned to @Frenzie earlier, I choose to change the code because it was the quickest for me to change. But does not reflect what will be merged in the future.
I'll make some changes in the documentation to explain more accurately how to use the extension. Change also the wording to explain that overriding those method might trigger some side-effects.

aledeg commented:
@Frenzie @kevinpapst @Alkarex I've changed the content of this PR. I've updated only the documentation to make it clearer (I hope!). Let me know what you think! Thanks

Frenzie commented:
Basically lgtm, but I'll go through with an actual editor's hat on tomorrow (please ping me if I forget).

aledeg commented:
I will! Thanks

kevinpapst reviewed (commented):
More documentation is always a great thing, I hope it leads to more new extensions 👍 
Just two change wishes from my side 😊

aledeg commented on docs/en/developers/03_Backend/05_Extensions.md:
```diff
@@ -279,33 +279,29 @@ A __system__ extension in comparison is enabled for every account.
+* the `registerController` method register an extension controller in FreshRSS. The selected controller must be defined in the extension _Controllers_ folder, its file name must be _<name>Controller.php_, and its class name must be *FreshExtension\_<name>\_Controller*.
+* the `registerViews` method registers the extension views in FreshRSS.
+* the `registerTranslates` method registers the extension translation files in FreshRSS.
+* the `registerHook` method registers hook actions in different part of the application.
```
Here is just the list of methods available you should not modify. The hook system is explained elsewhere. I think it is better to keep it as-is.

aledeg commented on docs/en/developers/03_Backend/05_Extensions.md:
```diff
@@ -279,33 +279,29 @@ A __system__ extension in comparison is enabled for every account.
-* `registerViews()`
-* `registerTranslates()`
-* `registerHook($hook_name, $hook_function)`
+> Note that if you modify the later set of methods, you might break the extension system. Thus making FreshRSS unusable. So it's highly recommended to let those unmodified.
```
The things is that you can modify them but it can be dangerous for the application. Those methods wont be removed.

kevinpapst replied:
huh? why did you made all the changes in the first place if you don't plan to fix the real problem? well, ok. 
it was just a comment what I would do... I am out, great pr!

aledeg replied:
I removed all the changes. This PR is only a documentation PR now to clarify things.
